Game logic

Maclenn77 opened the pull request:
We added the game logic and updated the main binary file

### New features:

- Create two Player objects at the beginning.
- Create a Match object after asking the names of players
- Indicate who start the match
- Indicate next turn
- Check all winner cases after the 5th turn
- Ask players if they want to play again
- Exit the game if they don't want to play the game or after writing 'exit' in a move
- Show the score of each player and total matches played

**Thanks for reviewing!**

RossielCS committed:
Add classes file

- Player class
- New players with name
- Remove .gitkeep

Maclenn77 committed:
Merge branch 'game_logic' of github.com:Maclenn77/Tic_tac_toe into game_logic

Maclenn77 committed:
Created Match class. Select random start.

RossielCS committed:
Add random select for players

- Add piece var to player, board with numbers

RossielCS committed:
Add place_move method

- var separator to const
- verify move method

Maclenn77 committed:
Changed array to hash. Now user input value and change board.

RossielCS committed:
Add conditionals in place_method

- It verify player's input

RossielCS committed:
Add players symbols to place_move

- Change symbol O to T for testing

Maclenn77 committed:
Refactored the function for moves

Maclenn77 committed:
Drafting next turn method

RossielCS committed:
Add next_turn method

- It automatically choses the next turn

Maclenn77 committed:
Refactored place_move method

Maclenn77 committed:
Created check_winner method. Lacking array for checking winner cases

RossielCS committed:
Add while loop to next_turn

- Loop for counting the turns
- Will exit the game

Maclenn77 committed:
Finished check_winner method

RossielCS committed:
Add keep_playing? method

- It calls restart_match method

Maclenn77 committed:
Creating a loop for restart_match

Maclenn77 committed:
Created loop for replaying. There's issues in the loop.

RossielCS committed:
change game method to Match class

- It's restart_match

RossielCS committed:
Add score method

- Count player's points

Maclenn77 committed:
Added scores and counting matches

RossielCS committed:
Remove main.rb in root

Maclenn77 committed:
Merge branch 'user_interface' of github.com:Maclenn77/Tic_tac_toe into game_logic

RossielCS committed:
Add hash for puts

- Add content in main_test and puts_hash

Maclenn77 committed:
Adding output at hash

Maclenn77 committed:
Taking changes of user interface

Merge branch 'develop' of github.com:Maclenn77/Tic_tac_toe into game_logic

RossielCS committed:
Move puts to puts_hash in main

- Add call method to hash item

Maclenn77 committed:
Added exit keyword. Stylized welcome message

RossielCS committed:
Move display_board content to main

- board_hash for rows and separator

Maclenn77 committed:
Removed message before saying that it's a draw

RossielCS committed:
Remove regex in keep_playin?

- Use equality operator
- Change play_again

Maclenn77 committed:
Removed global class variable. Changed like instance

euqueme commented on lib/classes.rb:
```diff
@@ -0,0 +1,130 @@
+    display_board(board_hash)
+  end
+
+  def next_turn(*args, puts_hash, board_hash)
+    @turn = 0
+    while @turn < 9
+      @turn += 1
+      if @turn.odd?
+        player_turn = args[0..-1].detect { |a| a.piece == 'X' }
+        player_next = args[0..-1].detect { |a| a.piece == 'O' }
+      else
+        player_turn = args[0..-1].detect { |a| a.piece == 'O' }
+        player_next = args[0..-1].detect { |a| a.piece == 'X' }
+      end
+      place_move(player_turn, puts_hash, board_hash)
+      check_winner(player_turn, player_next, puts_hash, board_hash) if @turn >= 5
+      puts_hash['your_turn'].call(player_next.name) if @turn < 9
+    end
+    puts_hash['draw'].call
+    keep_playing?(player_turn, player_next, puts_hash, board_hash)
+  end
```
* [x] No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- [x] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing).([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- This method has many responsibilities even though it should only change the player's turn it looks like it's handling the main gameplay placing moves, validating the winner, and also printing results on screen, the content of this whole method belongs in the loop that should exist in the main.rb file, so let's move it there.

euqueme commented on lib/classes.rb:
```diff
@@ -0,0 +1,130 @@
+    board_hash['row_three'].call(@hash)
+  end
+
+  def place_move(player_one, puts_hash, board_hash)
+    move = 0
+    until ((1..9).include? move) && (@hash.value? move)
+      puts_hash['another_num'].call
+      move = gets.chomp
+      exit if move == 'exit'
+      move = move.to_i
+    end
+    @hash[move] = player_one.piece
+    display_board(board_hash)
+  end
```
* [x] No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- [x] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing).([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- The responsibility of this method should be only placing a move on the board, instead, it has a loop that it's requesting for input and printing on screen, I think a great part of the behavior of this method also belongs in the main.rb file

euqueme commented on lib/classes.rb:
```diff
@@ -0,0 +1,130 @@
+    keep_playing?(player_turn, player_next, puts_hash, board_hash)
+  end
+
+  def check_winner(player_turn, player_next, puts_hash, board_hash)
+    winner = false
+    winner_cases = [[@hash[1], @hash[2], @hash[3]], [@hash[1], @hash[4], @hash[7]],
+                    [@hash[1], @hash[5], @hash[9]], [@hash[2], @hash[5], @hash[8]],
+                    [@hash[3], @hash[6], @hash[9]], [@hash[3], @hash[5], @hash[7]],
+                    [@hash[4], @hash[5], @hash[6]], [@hash[7], @hash[8], @hash[9]]]
+    winner_cases.each { |cases| winner = true if cases.all?(player_turn.piece) }
+    if winner
+      puts_hash['you_won'].call(player_turn.name)
+      player_turn.change_score
+      keep_playing?(player_turn, player_next, puts_hash, board_hash)
+    end
+    winner
+  end
```
* [x] No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- [ ] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing).([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- This method returns a boolean which is great you could use it to print the message in your main.rb file instead of printing it here so lines 85 to 89 belong in the main.rb file

euqueme commented on lib/classes.rb:
```diff
@@ -0,0 +1,130 @@
+    winner
+  end
+
+  def keep_playing?(player_one, player_two, puts_hash, board_hash)
+    condition = false
+    puts_hash['play_again'].call
+    choice = gets.chomp
+    until condition
+      if choice == 'y'
+        condition = true
+        restart_match(player_one, player_two, puts_hash, board_hash)
+      elsif choice == 'n'
+        puts_hash['exit_match'].call
+        total_score(player_one, player_two)
+        exit
+      else
+        puts_hash['valid_op'].call
+        choice = gets.chomp
+      end
+    end
+  end
```
* [x] No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- [x] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing).([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- a great part of this method also belongs in the main.rb file no printing or gets.chomp should be here, this method should only return a boolean that decides to start another match or not

euqueme commented on lib/classes.rb:
```diff
@@ -0,0 +1,130 @@
+    end
+  end
+
+  def total_score(player_one, player_two, matches)
+    puts "Total matches: #{matches} Total Score:\n
+          #{player_one.name}: #{player_one.score}\n
+          #{player_two.name}: #{player_two.score}"
+  end
```
* [x] No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- this method it's printing on screen this should only happen in the main.rb file so this method could actually just return a string and then inside your main.rb file you could call it like this:

```ruby
puts total_score(player_one, player_two, matches)
```

euqueme commented on lib/classes.rb:
```diff
@@ -0,0 +1,130 @@
+          #{player_two.name}: #{player_two.score}"
+  end
+
+  def restart_match(player_one, player_two, puts_hash, board_hash)
+    if @matches.positive?
+      total_score(player_one, player_two, @matches)
+      puts_hash['yes_play'].call
+    end
+    matches = @matches + 1
+    match = Match.new(matches)
+    first = match.who_is_first?(player_one, player_two)
+    puts_hash['first_move'].call(first)
+    match.display_board(board_hash)
+    match.next_turn(player_one, player_two, puts_hash, board_hash)
+  end
```
* [x] No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- [x] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing).([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- restart_match should only return a boolean that validates an input and returns true or false but instead, it's validating, displaying the board, and displaying messages on the screen it has too many responsibilities so most of the behavior of this method belongs in the main.rb file

euqueme reviewed (changes requested):
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @Maclenn77 and @RossielCS  :balloon: :lollipop: 

### You've done a great job with this project, I like when games ask for the players names I think this feature makes it more customizable :video_game: and you should feel very proud of your work :star2: :1st_place_medal: there is still some work to do on your code :construction: :construction_worker: :construction_worker_woman:  so please keep reading

#### Behaviour

- [x] your game it's letting players have empty names, let's add a validation to prevent this from happening

![image](https://user-images.githubusercontent.com/52209067/80563974-d38e0600-89a9-11ea-8ecb-3d1bc3cad2b9.png)

- [x] the game is showing these errors when exiting let's look into them :mag: 

![image](https://user-images.githubusercontent.com/52209067/80564093-1b149200-89aa-11ea-9f91-b8d6827689b8.png)

#### Implementation details
* [x] No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- [x] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing).([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- Even though putting behavior on methods makes your code look organized it doesn't mean it's a good practice, methods should always have just one responsibility and in this particular project, it is requested that printing and requesting for input only happen in the main.rb file, the reason we are so strict with this rule is that later on (in a future project) you'll have to build test cases of this project and it's more difficult to build test cases of methods that have many responsibilities.

- That being said, I have left comments on your code where changes need to be done to respect both of the above rules

[OPTIONAL]

- [ ] This is not a mandatory requirement yet but I strongly recommend to put your Player and Match classes on separate files within the lib directory.

### That's it, team! :raised_hands: 

### Please :pray: do the changes :technologist: :woman_technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any questions? I'd be happy to answer in slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :traffic_light:

RossielCS committed:
Add new methods

- change_name in main
- valid_name, write_name in classes

Maclenn77 committed:
Added who_is_first and display_boar in main

RossielCS committed:
Add method to main

- valid_move checks input

RossielCS committed:
Add loop_match method

- It keeps a loop for the game

Maclenn77 committed:
Added check_winner to loop_match. Doing a loop for keep playing

RossielCS committed:
Add display_result method

- Remove comments and methods

RossielCS committed:
Create files for each class

- new files in lib

Maclenn77 commented:
Thanks for reviewing, @euqueme !

### Changes done:
- Modified all the methods with gets.chomp or put
- Added all the puts and gets.chomp in main
- Solved the bug when the user says 'no' to keep playing
- Simplified methods, so they will do only one thing at a time

**Have a good day!**

pelzolga123 reviewed (approved):
Approved.

pelzolga123 commented:
## Hi Team,
## Project looks amazing. Congratulation :+1: Approved.
Go ahead and merge to develop.
_Goodluck in your next milestone_ *Happy coding*
